### PR TITLE
HTCONDOR-2523 Replace ClassAdListDoesNotDelete ads with vector

### PR DESCRIPTION
**Pull request by GregThain**
<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [ ] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [ ] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [ ] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed  (documentation [build logs](https://readthedocs.org/projects/htcondor/builds/))
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [ ] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

**Comment by GregThain** [comment]
`ClassAdListDoesNotDeleteAds` also has an internal `HashTable` to silently reject insertion of the same ad address more than once.  I can't see anywhere in the code where we try to do that, so this PR doesn't replicate that functionality.  Something to keep an eye on.

File: src/condor_negotiator.V6/matchmaker.cpp [hunk 1]
```diff
@@ -1625,11 +1624,25 @@ int count_effective_slots(ClassAdListDoesNotDeleteAds& startdAds, ExprTree* cons
 }
 
 
+static int 
+CountMatches(std::vector<ClassAd *> ads, classad::ExprTree* constraint) {
```
Comment (JaimeFrey) [inline src/condor_negotiator.V6/matchmaker.cpp:1628]: Shouldn't this be pass-by-reference for the ads vector?

File: src/condor_negotiator.V6/matchmaker.cpp [hunk 2]
```diff
@@ -2747,43 +2758,16 @@ trimStartdAds_PreemptionLogic(ClassAdListDoesNotDeleteAds &startdAds) const
...
-				removed++;
-			}
-		}
-	}
-	startdAds.Close();
-
-	dprintf(D_FULLDEBUG,
-			"Trimmed out %d startd ads due to NEGOTIATOR_CONSIDER_PREEMPTION=False\n",
-			removed);
-
```
Comment (JaimeFrey) [inline src/condor_negotiator.V6/matchmaker.cpp:2772]: Why did you remove this block? It's how we ignore claimed slots when preemption is disabled.

**Review by JaimeFrey — CHANGES_REQUESTED** [review]

File: src/condor_negotiator.V6/matchmaker.cpp (hunk 1 above)
Comment (GregThain) [reply src/condor_negotiator.V6/matchmaker.cpp:1628]: Yes.

**Review by JaimeFrey — APPROVED** [review]
LGTM
